File: src/vm/lexer.c
#include <stdint.h>
#include <stdio.h>
#include <string.h>

#include "common.h"
#include "debug.h"
#include "vm/lexer.h"
#include "vm/token_type.h"
#include "vm/values.h"

lexer _lexer; /* Static instance of lexer. Rewrite to use pointers instead. */

void init_lexer(string source) {
    _lexer.left = source;
    _lexer.right = source;
    _lexer.line = 1;
}

/**
 * Compares provided input against the expected token type and returns the type if they're equal,
 * otherwise returns a identifier token
 * @param start the starting index of the string
 * @param length the length of the remaining string after the starting character
 * @param remainder the substring of the keyword excluding the start character
 * @param type the expected token_type
 * 
 * @returns The expected token type if memcmp returns 1
*/
static token_type check_keyword(uint8_t start, uint8_t length, string remainder,
                                token_type type) {
    if ((_lexer.right - _lexer.left == start + length) &&
        memcmp(_lexer.left + start, remainder, length) == 0)
        return type;

    return TOKEN_IDENTIFIER;
}

static char peek() { return *_lexer.right; }

/*static char peek_next() { return _lexer.right[1]; }*/

static bool eof() { return *_lexer.right == '\0'; }

static char advance() {
    _lexer.right++;
    return _lexer.right[-1];
}

static void skip_whitespace() {
    for (;;) {
    char c = peek();
    switch (c) {
      case ' ':
      case '\r':
      case '\t':
        advance();
        break;
      default:
        return;
    }
  }
}

static token make_token(token_type type) {
    token tkn;
    tkn.type = type;
    tkn.line = _lexer.line;
    tkn.length = (int)(_lexer.right - _lexer.left);
    tkn.start = _lexer.left;
    return tkn;
}

static token error_token(string message) {
    token tkn;
    tkn.type = TOKEN_ERROR;
    tkn.start = _lexer.left;
    tkn.length = strlen(message);
    tkn.line = _lexer.line;
    return tkn;
}

static bool match(char expected) {
    if (eof())
        return false;
    if (*_lexer.right != expected)
        return false;
    _lexer.right++;
    return true;
}

static token str() {
    while (peek() != '"' && !eof()) {
        if (peek() == '\n')
            _lexer.line++;
        advance();
    }

    if (eof())
        return error_token("Unterminated string");

    advance(); // Eat the closing quotation

    return make_token(TOKEN_STRING);
}

static bool is_digit(char c) { return c >= '0' && c <= '9'; }

static bool is_alpha(char c) {
    return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c == '_');
}

static token num() {
    while (is_digit(peek()) || peek() == '.')
        advance();

    return make_token(TOKEN_NUMBER);
}

static token ident() {
    while (is_alpha(peek()) || is_digit(peek()))
        advance();

    token_type type;

    switch (_lexer.left[0]) {
    case 'a':
        type = check_keyword(1, 2, "nd", TOKEN_AND);
        break;
    case 'c':
        type = check_keyword(1, 4, "lass", TOKEN_CLASS);
        break;
    case 'e':
        type = check_keyword(1, 3, "lse", TOKEN_ELSE);
        break;
    case 'f':
        if (_lexer.right - _lexer.left > 1) {
            switch (_lexer.left[1]) {
            case 'a':
                type = check_keyword(2, 3, "lse", TOKEN_FALSE);
                break;
            case 'o':
                break;
            case 'u':
                break;
            }
        }
        break;
    case 'i':
        type = check_keyword(1, 1, "f", TOKEN_IF);
        break;
    case 'n':
        if (_lexer.right - _lexer.left > 1)
            switch (_lexer.left[1]) {
            case 'o':
                type = check_keyword(2, 2, "ne", TOKEN_NONE);
                break;
            case 't':
                type = check_keyword(2, 1, "t", TOKEN_NOT);
                break;
            }
        break;
    case 'o':
        type = check_keyword(1, 1, "r", TOKEN_OR);
        break;
    case 'p':
        if (_lexer.right - _lexer.left > 2) {
            switch (_lexer.left[2]) {
            case 'i':
                type = check_keyword(3, 2, "nt", TOKEN_PRINT);
                break;
            case 'o':
                type = check_keyword(3, 1, "c", TOKEN_PROC);
                break;
            }
        }
        break;
    case 'r':
        type = check_keyword(1, 5, "eturn", TOKEN_RETURN);
        break;
    case 's':
        type = check_keyword(1, 4, "uper", TOKEN_SUPER);
        break;
    case 'l':
        type = check_keyword(1, 2, "et", TOKEN_LET);
        break;
    case 'w':
        type = check_keyword(1, 4, "hile", TOKEN_WHILE);
        break;
    case 't':
        type = check_keyword(1, 3, "rue", TOKEN_TRUE);
        break;
    default:
        type = TOKEN_IDENTIFIER;
        break;
    }

    return make_token(type);
}

token scan_token() {
    skip_whitespace();

    _lexer.left = _lexer.right;

    if (eof())
        return make_token(TOKEN_EOF);

    char c = advance();

    if(is_alpha(c)) return ident();
    if(is_digit(c)) return num();

    switch (c) {
    case '(':
        return make_token(TOKEN_LEFT_PAREN);
    case ')':
        return make_token(TOKEN_RIGHT_PAREN);
    case '{':
        return make_token(TOKEN_LEFT_BRACE);
    case '}':
        return make_token(TOKEN_RIGHT_BRACE);
    case ';':
        return make_token(TOKEN_SEMICOLON);
    case ',':
        return make_token(TOKEN_COMMA);
    case '.':
        return make_token(TOKEN_DOT);
    case '-':
        return make_token(TOKEN_MINUS);
    case '+':
        return make_token(TOKEN_PLUS);
    case '/':
        return make_token(TOKEN_SLASH);
    case '*':
        return make_token(TOKEN_STAR);
    case '"':
        return str();
    case '!':
        return make_token(match('=') ? TOKEN_BANG_EQUAL : TOKEN_BANG);
    case '=':
        return make_token(match('=') ? TOKEN_EQUAL_EQUAL : TOKEN_EQUAL);
    case '<':
        return make_token(match('=') ? TOKEN_LESS_EQUAL : TOKEN_LESS);
    case '>':
        return make_token(match('=') ? TOKEN_GREATER_EQUAL : TOKEN_GREATER);
    case '#':
        while (peek() != '\n' && !eof())
            advance();
        return make_token(TOKEN_POUND);
    }

    return error_token("Unexpected token");
}

File: src/lexer.rs
use crate::tokens::*;
use crate::error::{parse_token_error, operation_error};

pub struct Lexer {
    source: String,
    line: i64,
    current: i64,
    start: i64,
    ch: char,
}

impl Lexer {
    pub fn new(source: String) -> Self {
        Lexer {
            source: source,
            line: 1,
            current: 0,
            start: 0,
            ch: '0',
        }
    }

    pub fn tokenize(&mut self) -> Vec<Token> {
        let mut tokens: Vec<Token> = vec![];

        while !self.eof() {
            tokens.push(self.next_token())
        }

        tokens.push(self.make_token(TokenType::TokenEof));
        tokens
    }

    fn next_token(&mut self) -> Token {
        self.skip_whitespace();

        self.read_char();
        
        if self.eof() {
            return self.make_token(TokenType::TokenEof);
        }

        if self.ch.is_alphabetic() {
            return self.read_ident()
        }

        if self.ch.is_numeric() {
            return self.read_number()
        }

        if self.ch == '"' {
            return self.read_string()
        }

        match self.ch {
            '(' => return self.make_token(TokenType::TokenLeftParen),
            ')' => return self.make_token(TokenType::TokenRightParen),
            '{' => return self.make_token(TokenType::TokenLeftBrace),
            '}' => return self.make_token(TokenType::TokenRightBrace),
            '[' => return self.make_token(TokenType::TokenLeftBracket),
            ']' => return self.make_token(TokenType::TokenRightBracket),
            '.' => return self.make_token(TokenType::TokenDot),
            ',' => return self.make_token(TokenType::TokenComma),
            ';' => return self.make_token(TokenType::TokenSemicolon),
            '+' => return self.make_token(TokenType::TokenPlus),
            '-' => return self.make_token(TokenType::TokenMinus),
            '*' => return self.make_token(TokenType::TokenStar),
            '/' => return self.make_token(TokenType::TokenSlash),
            '^' => return self.make_token(TokenType::TokenPower),
            '%' => return self.make_token(TokenType::TokenModulo),
            '=' => {
                match self.next_char('=') {
                    true => return self.make_token(TokenType::TokenEqualEqual),
                    false => return self.make_token(TokenType::TokenEqual),
                };
            },
            '!' => {
                match self.next_char('=') {
                    true => return self.make_token(TokenType::TokenBangEqual),
                    false => return self.make_token(TokenType::TokenBang),
                }
            },
            '>' => {
                match self.next_char('=') {
                    true => return self.make_token(TokenType::TokenGreaterEqual),
                    false => return self.make_token(TokenType::TokenGreater),
                }
            },
            '<' => {
                match self.next_char('=') {
                    true => return self.make_token(TokenType::TokenLessEqual),
                    false => return self.make_token(TokenType::TokenLess),
                }
            },
            '#' => return self.make_token(TokenType::TokenPound),
            '&' => {
                match self.next_char('&') {
                    true => return self.make_token(TokenType::TokenAnd),
                    false => operation_error("Invalid AND operation. Use && instead.", &self.line.to_string()),
                }
            }
            '|' => {
                match self.next_char('|') {
                    true => return self.make_token(TokenType::TokenOr),
                    false => operation_error("Invalid OR operation. Use || instead.", &self.line.to_string()),
                }
            }
            _ => parse_token_error("Invalid character", &self.ch.to_string(), &self.line.to_string()),
        };

        unreachable!()
    }
    
    fn read_char(&mut self) {
        let index: usize = self.current.try_into().unwrap();
        self.ch = self.source.as_bytes()[index] as char;
        self.current += 1;
    }

    fn make_token(&mut self, token_type: TokenType) -> Token {
        #[allow(unused_assignments)]
        let mut literal = "";            

        match token_type {
            TokenType::TokenLeftParen => literal = "(",
            TokenType::TokenRightParen => literal = ")",
            TokenType::TokenLeftBrace => literal = "{",
            TokenType::TokenRightBrace => literal = "}",
            TokenType::TokenLeftBracket => literal = "[",
            TokenType::TokenRightBracket => literal = "]",
            TokenType::TokenDot => literal = ".",
            TokenType::TokenComma => literal = ",",
            TokenType::TokenSemicolon => literal = ";",
            TokenType::TokenPlus => literal = "+",
            TokenType::TokenMinus => literal = "-",
            TokenType::TokenStar => literal = "*",
            TokenType::TokenSlash => literal = "/",
            TokenType::TokenPower => literal = "^",
            TokenType::TokenModulo => literal = "%",
            TokenType::TokenPound => literal = "#",
            TokenType::TokenEqual => literal = "=",
            TokenType::TokenEqualEqual => literal = "==",
            TokenType::TokenBang => literal = "!",
            TokenType::TokenBangEqual => literal = "!=",
            TokenType::TokenLess => literal = "<",
            TokenType::TokenLessEqual => literal = "<=",
            TokenType::TokenGreater => literal = ">",
            TokenType::TokenGreaterEqual => literal = ">=",
            TokenType::TokenAnd => literal = "&&",
            TokenType::TokenOr => literal = "||",
            TokenType::TokenProc => literal = "proc",
            TokenType::TokenIf => literal = "if",
            TokenType::TokenElse => literal = "else",
            TokenType::TokenFor => literal = "for",
            TokenType::TokenTrue => literal = "true",
            TokenType::TokenFalse => literal = "false",
            TokenType::TokenLet => literal = "let",
            TokenType::TokenNull => literal = "null",
            TokenType::TokenNot => literal = "not",
            TokenType::TokenReturn => literal = "return",
            TokenType::TokenEof => literal = "eof",
            _ => literal = "error",            
        }

        Token {
            token_type: token_type,
            line: self.line,
            literal: String::from(literal),
        }
    }
    
    fn peek(&mut self) -> char {
        if self.eof() {
            return '\0'
        }
        let index: usize = self.current.try_into().unwrap();
        self.source.as_bytes()[index] as char
    }

    fn next_char(&mut self, ch: char) -> bool {
        if self.eof() {
            return false
        }

        let index: usize = self.current.try_into().unwrap();
        if self.source.as_bytes()[index] as char == ch {
            self.ch = ch;
            self.current += 1;
            return true
        }

        false
    }
    
    fn read_ident(&mut self) -> Token {
        self.start = self.current - 1;
        while self.peek().is_alphabetic() {
            self.read_char();
        }

        let ident: String = self.source
            .chars()
            .skip(self.start as usize)
            .take((self.current-self.start) as usize)
            .collect();

    
        let token_type = reserved_keyword(&ident.to_lowercase());

        if token_type != TokenType::TokenIdentifier {
            return self.make_token(token_type)
        }

        Token {
            line: self.line,
            literal: ident,
            token_type: token_type
        }
    }
    
    fn read_number(&mut self) -> Token {
        self.start = self.current - 1;

        while self.peek().is_numeric() || self.peek() == '.' {
            self.read_char();
        }

        let ident: String = self.source
            .chars()
            .skip(self.start as usize)
            .take((self.current-self.start) as usize)
            .collect();

        Token {
            line: self.line,
            literal: ident,
            token_type: TokenType::TokenNumber,
        }
    }

    fn read_string(&mut self) -> Token {
        // Eat the first quotation mark
        self.read_char();

        self.start = self.current - 1;
        
        while self.peek() != '"' {
            self.read_char();
        }

        // Eat the last quotation mark
        self.read_char();

        let ident: String = self.source
            .chars()
            .skip(self.start as usize)
            .take(((self.current-self.start) - 1) as usize)
            .collect();

        Token {
            line: self.line,
            literal: ident,
            token_type: TokenType::TokenString,
        }
    }
    
    fn skip_whitespace(&mut self) {
        loop {
            match self.peek() {
                ' ' => self.read_char(),
                '\t' => { 
                    self.read_char() 
                }
                '\r' => {
                    self.read_char() 
                }
                '\n' => { 
                    self.line += 1; 
                    self.read_char() 
                }
                _ => return,
            };
        }
    }

    fn eof(&self) -> bool {
        self.ch == '\0' || self.current == self.source.len() as i64
    }
}
